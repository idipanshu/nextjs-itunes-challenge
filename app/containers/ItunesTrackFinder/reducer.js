import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: searchContainerTypes, Creators: searchContainerCreators } = createActions({
  requestGetSongs: ['searchedTerm'],
  requestGetSongDetails: ['trackId'],
  successGetSongs: ['data'],
  successGetSongDetails: ['data'],
  failureGetSongs: ['error'],
  clearSongs: []
});

export const initialState = { searchedTerm: null, trackId: null, songsData: {}, trackDetails: {}, songsError: null };

/* eslint-disable default-case, no-param-reassign */
export const searchContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case searchContainerTypes.REQUEST_GET_SONGS:
        draft.searchedTerm = action.searchedTerm;
        break;

      case searchContainerTypes.CLEAR_SONGS:
        draft.searchedTerm = null;
        draft.songsData = {};
        draft.songsError = null;
        break;

      case searchContainerTypes.SUCCESS_GET_SONGS:
        draft.songsData = action.data;
        break;

      case searchContainerTypes.REQUEST_GET_SONG_DETAILS:
        draft.trackId = action.trackId;
        break;

      case searchContainerTypes.CLEAR_SONG_DETAILS:
        draft.trackId = null;
        draft.trackDetails = {};
        draft.songsError = null;
        break;

      case searchContainerTypes.SUCCESS_GET_SONG_DETAILS:
        draft.trackDetails = action.data;
        break;

      case searchContainerTypes.FAILURE_GET_SONGS:
        draft.songsError = get(action.error, 'message', 'something_went_wrong');
        break;
    }
  });

export default searchContainerReducer;
