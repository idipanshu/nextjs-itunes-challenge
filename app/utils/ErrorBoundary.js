import React, { Component } from 'react';

export default class ErrorBoundary extends Component {
  constructor(props) {
    super(props);

    this.state = {
      error: false
    };
  }

  static getDerivedStateFromError(error) {
    if (error) {
      return { error: true };
    }
  }

  componentDidCatch(error) {
    throw Error(error);
  }

  render() {
    if (this.state.error) {
      return (
        <div>
          <h1>Something went wrong.</h1>
        </div>
      );
    }

    return this.props.children;
  }
}
