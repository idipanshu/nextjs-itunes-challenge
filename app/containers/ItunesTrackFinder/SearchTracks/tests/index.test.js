import { fireEvent } from '@testing-library/dom';
import { renderProvider, timeout } from '@app/utils/testUtils';

import { searchContainerTypes } from '../../reducer';
import { SearchTracksTests as SearchTracks, mapDispatchToProps } from '../index';

describe('<SeachContainer /> test', () => {
  let submitSpy;

  beforeEach(() => {
    submitSpy = jest.fn();
  });

  it('should render and match the snapshot', () => {
    const clearDispatch = jest.fn();

    const { baseElement } = renderProvider(
      <SearchTracks dispatchClearItunesTracks={clearDispatch} dispatchGetItunesTracks={submitSpy} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should call dispatchClearItunesTracks on empty change', async () => {
    const getSongsSpy = jest.fn();
    const clearItunesTracksSpy = jest.fn();
    const { getByTestId } = renderProvider(
      <SearchTracks dispatchClearItunesTracks={clearItunesTracksSpy} dispatchGetItunesTracks={getSongsSpy} />
    );
    fireEvent.change(getByTestId('search-input'), {
      target: { value: 'test' }
    });
    await timeout(500);
    expect(getSongsSpy).toBeCalled();

    fireEvent.change(getByTestId('search-input'), {
      target: { value: '' }
    });
    await timeout(500);
    expect(clearItunesTracksSpy).toBeCalled();
  });

  it('should call dispatchGetItunesTracks on when some input is provided', async () => {
    const searchedTerm = 'find-tracks-on-itunes';
    const clearDispatch = jest.fn();
    const { getByTestId } = renderProvider(
      <SearchTracks dispatchClearItunesTracks={clearDispatch} dispatchGetItunesTracks={submitSpy} />
    );
    const searchBar = getByTestId('search-input');

    fireEvent.change(searchBar, {
      target: { value: searchedTerm }
    });
    await timeout(500);

    expect(submitSpy).toBeCalledWith(searchedTerm);

    fireEvent.keyDown(searchBar, {
      key: 'Enter',
      code: 13,
      charCode: 13
    });
    expect(submitSpy).toBeCalledWith(searchedTerm);
  });

  it('should dispatchGetItunesTracks on update if searchTerm already exists', async () => {
    const searchedTerm = 'arijit';
    const clearDispatchSpy = jest.fn();
    renderProvider(
      <SearchTracks
        searchedTerm={searchedTerm}
        songsData={[]}
        dispatchGetItunesTracks={submitSpy}
        dispatchClearItunesTracks={clearDispatchSpy}
      />
    );

    await timeout(500);

    expect(submitSpy).toBeCalledTimes(0);
  });

  it('should validate mapDispatchToProps actions', async () => {
    const dispatchSongsDataSpy = jest.fn();
    const searchedTerm = 'arijit';
    const actions = {
      dispatchGetItunesTracks: { searchedTerm, type: searchContainerTypes.REQUEST_GET_SONGS },
      dispatchClearItunesTracks: { type: searchContainerTypes.CLEAR_SONGS }
    };

    const props = mapDispatchToProps(dispatchSongsDataSpy);
    props.dispatchGetItunesTracks(searchedTerm);
    expect(dispatchSongsDataSpy).toHaveBeenCalledWith(actions.dispatchGetItunesTracks);

    await timeout(500);
    props.dispatchClearItunesTracks();
    expect(dispatchSongsDataSpy).toHaveBeenCalledWith(actions.dispatchClearItunesTracks);
  });

  it('should render same number of cards as results', () => {
    const songsData = [
      {
        trackId: 741296001,
        trackName: 'Arijit',
        artistName: 'Test',
        currency: 'USD'
      },
      {
        trackId: 741297835,
        trackName: 'Arijit',
        artistName: 'John Doe',
        currency: 'INR'
      }
    ];

    const submitStub = jest.fn();
    const { getAllByTestId } = renderProvider(
      <SearchTracks songsData={songsData} dispatchClearItunesTracks={submitStub} dispatchGetItunesTracks={submitSpy} />
    );
    expect(getAllByTestId('music-card').length).toBe(songsData.length);
  });
});
