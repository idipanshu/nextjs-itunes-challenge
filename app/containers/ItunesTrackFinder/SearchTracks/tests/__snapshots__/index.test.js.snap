// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SeachContainer /> test should render and match the snapshot 1`] = `
.c0 {
  max-width: 85%;
  margin: 1rem auto;
  padding: 0.5rem;
}

.c1.c1 {
  padding: 0.1rem;
  margin: 1rem 0;
  background-color: #f8f8f8;
}

.c2.c2 {
  padding: 0.5rem 1rem;
  font-size: 1.3rem;
  color: #020826;
}

<body>
  <div>
    <div
      class="c0"
    >
      <div
        class="ant-card ant-card-bordered c1"
      >
        <div
          class="ant-card-body"
        >
          <input
            class="ant-input c2"
            data-testid="search-input"
            placeholder="Enter name of a song or an artist"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
